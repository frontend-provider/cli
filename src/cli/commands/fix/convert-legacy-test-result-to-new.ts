import { DepGraphData } from '@snyk/dep-graph';
import { TestResult } from '../../../lib/ecosystems/types';
import { TestResult as LegacyTestResult } from '../../../lib/snyk-test/legacy';

export function convertLegacyTestResultToNew(
  testResult: LegacyTestResult,
): TestResult {
  return {
    issuesData: {} as any, // TODO: add converter
    issues: [], // TODO: add converter
    remediation: testResult.remediation,
    // TODO: grab this once Ecosystems flow starts sending back ScanResult
    depGraphData: {} as DepGraphData,
  };
}
