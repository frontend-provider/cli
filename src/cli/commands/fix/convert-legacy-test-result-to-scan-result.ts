import { ScanResult } from '../../../lib/ecosystems/types';
import { TestResult } from '../../../lib/snyk-test/legacy';

export function convertLegacyTestResultToScanResult(
  testResult: TestResult,
): ScanResult {
  if (!testResult.packageManager) {
    throw new Error(
      'Only results with packageManagers are supported for conversion',
    );
  }
  return {
    identity: {
      type: testResult.packageManager,
      // this is because not all plugins send it back today, but we should always have it
      targetFile: testResult.targetFile || testResult.displayTargetFile,
    },
    name: testResult.projectName,
    // TODO: grab this once Ecosystems flow starts sending back ScanResult
    facts: [],
    policy: testResult.policy,
    // TODO: grab this once Ecosystems flow starts sending back ScanResult
    target: {} as any,
  };
}
