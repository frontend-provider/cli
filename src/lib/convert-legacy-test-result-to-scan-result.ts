import { ScanResult } from 'snyk-cpp-plugin';
import { TestResult } from './snyk-test/legacy';

export function convertLegacyTestResultToScanResult(
  testResult: TestResult,
): ScanResult {
  if (!testResult.packageManager) {
    throw new Error(
      'Only results with packageManagers are supported for conversion',
    );
  }
  return {
    identity: {
      type: testResult.packageManager,
      // this is because not all plugins send it back today, but we should always have it
      targetFile: testResult.targetFile || testResult.displayTargetFile,
    },
    name: testResult.projectName, // TODO: confirm
    facts: [
      // TODO: this needs to come from project Scan result
    ],
    policy: testResult.policy,
    target: {} as any, // TODO: this needs to come from project Scan result
  };
}
