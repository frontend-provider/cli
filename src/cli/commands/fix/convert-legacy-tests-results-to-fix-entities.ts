import * as fs from 'fs';
import * as pathLib from 'path';
import { convertLegacyTestResultToNew } from './convert-legacy-test-result-to-new';
import { convertLegacyTestResultToScanResult } from './convert-legacy-test-result-to-scan-result';
import { TestResult } from '../../../lib/snyk-test/legacy';
import { EntityToFix } from '@snyk/fix';
import { Options, TestOptions } from '../../../lib/types';

export function convertLegacyTestResultToFixEntities(
  testResults: (TestResult | TestResult[]) | Error,
  root: string,
  options: Partial<Options & TestOptions>,
): EntityToFix[] {
  if (testResults instanceof Error) {
    return [];
  }
  const oldResults = Array.isArray(testResults) ? testResults : [testResults];
  return oldResults.map((res) => ({
    options,
    workspace: {
      path: root,
      readFile: async (path: string) => {
        return fs.readFileSync(pathLib.resolve(root, path), 'utf8');
      },
      writeFile: async (path: string, content: string) => {
        return fs.writeFileSync(pathLib.resolve(root, path), content, 'utf8');
      },
    },
    scanResult: convertLegacyTestResultToScanResult(res),
    testResult: convertLegacyTestResultToNew(res),
  }));
}
