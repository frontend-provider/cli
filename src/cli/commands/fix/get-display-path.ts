import * as pathLib from 'path';

import { isLocalFolder } from '../../../lib/detect';

export function getDisplayPath(path: string): string {
  if (!isLocalFolder(path)) {
    return path;
  }
  if (path === process.cwd()) {
    return pathLib.parse(path).name;
  }
  return pathLib.relative(process.cwd(), path);
}
