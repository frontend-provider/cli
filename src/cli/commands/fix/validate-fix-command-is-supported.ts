import * as Debug from 'debug';

import { getEcosystemForTest } from '../../../lib/ecosystems';

import { isFeatureFlagSupportedForOrg } from '../../../lib/feature-flags';
import { CommandNotSupportedError } from '../../../lib/errors/command-not-supported';
import { FeatureNotSupportedByEcosystemError } from '../../../lib/errors/not-supported-by-ecosystem';
import { Options, TestOptions } from '../../../lib/types';
import { AuthFailedError } from '../../../lib/errors';

const debug = Debug('snyk-fix');
const snykFixFeatureFlag = 'cliSnykFix';

export async function validateFixCommandIsSupported(
  options: Options & TestOptions,
): Promise<boolean> {
  if (options.docker) {
    throw new FeatureNotSupportedByEcosystemError('snyk fix', 'docker');
  }

  const ecosystem = getEcosystemForTest(options);
  if (ecosystem) {
    throw new FeatureNotSupportedByEcosystemError('snyk fix', ecosystem);
  }

  const snykFixSupported = await isFeatureFlagSupportedForOrg(
    snykFixFeatureFlag,
    options.org,
  );

  debug('Feature flag check returned: ', snykFixSupported);

  if (snykFixSupported.code === 401 || snykFixSupported.code === 403) {
    throw AuthFailedError(snykFixSupported.error, snykFixSupported.code);
  }

  if (!snykFixSupported.ok) {
    throw new CommandNotSupportedError('snyk fix', options.org || undefined);
  }

  return true;
}
