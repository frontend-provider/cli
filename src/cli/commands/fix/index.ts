export = fix;

import * as Debug from 'debug';
import * as snykFix from '@snyk/fix';

import { MethodArgs } from '../../args';
import * as snyk from '../../../lib';
import { TestResult } from '../../../lib/snyk-test/legacy';

import { convertLegacyTestResultToFixEntities } from './convert-legacy-tests-results-to-fix-entities';
import { formatTestError } from '../test/format-test-error';
import { processCommandArgs } from '../process-command-args';
import { validateCredentials } from '../test/validate-credentials';
import { validateTestOptions } from '../test/validate-test-options';
import { setDefaultTestOptions } from '../test/set-default-test-options';
import { validateFixCommandIsSupported } from './validate-fix-command-is-supported';

const debug = Debug('snyk-fix');
const snykFixFeatureFlag = 'cliSnykFix';

async function fix(...args: MethodArgs): Promise<string> {
  const { options: rawOptions, paths } = await processCommandArgs(...args);
  const options = setDefaultTestOptions(rawOptions);
  await validateFixCommandIsSupported(options);
  validateTestOptions(options);
  validateCredentials(options);

  const results: snykFix.EntityToFix[] = [];
  results.push(...(await runSnykTestLegacy(options, paths)));

  // fix
  debug(
    `Organization has ${snykFixFeatureFlag} feature flag enabled for experimental Snyk fix functionality`,
  );
  const { fixSummary, meta } = await snykFix.fix(results);

  if (meta.fixed === 0) {
    throw new Error(fixSummary);
  }
  return fixSummary;
}

/* @deprecated
 * TODO: once project envelope is default all code below will be deleted
 * we should be calling test via new Ecosystems instead
 */
async function runSnykTestLegacy(
  options,
  paths,
): Promise<snykFix.EntityToFix[]> {
  const results: snykFix.EntityToFix[] = [];
  for (const path of paths) {
    // Create a copy of the options so a specific test can
    // modify them i.e. add `options.file` etc. We'll need
    // these options later.
    const snykTestOptions = {
      ...options,
      path,
      projectName: options['project-name'],
    };

    let testResults: TestResult | TestResult[];

    try {
      testResults = await snyk.test(path, snykTestOptions);
    } catch (error) {
      const testError = formatTestError(error);
      throw testError;
    }
    const resArray = Array.isArray(testResults) ? testResults : [testResults];
    const newRes = convertLegacyTestResultToFixEntities(resArray, path);
    results.push(...newRes);
  }
  return results;
}
