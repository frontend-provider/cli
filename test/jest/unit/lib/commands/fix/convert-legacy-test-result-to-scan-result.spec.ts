import * as fs from 'fs';
import * as path from 'path';

import { convertLegacyTestResultToScanResult } from '../../../../../../src/cli/commands/fix/convert-legacy-test-result-to-scan-result';

describe('Convert legacy TestResult to ScanResult', () => {
  it('can convert npm test result with no remediation', () => {
    const noRemediationRes = JSON.parse(
      fs.readFileSync(
        path.resolve(
          __dirname,
          '../../../../../',
          'acceptance/fixtures/npm-package-with-severity-override/test-graph-result-no-remediation.json',
        ),
        'utf8',
      ),
    );
    const res = convertLegacyTestResultToScanResult(noRemediationRes);
    expect(res).toMatchSnapshot();
  });

  it('can convert npm test result with remediation', () => {
    const withRemediation = JSON.parse(
      fs.readFileSync(
        path.resolve(
          __dirname,
          '../../../../../',
          'acceptance/fixtures/npm-package-with-severity-override/test-graph-result-patches.json',
        ),
        'utf8',
      ),
    );
    const res = convertLegacyTestResultToScanResult(withRemediation);
    expect(res).toMatchSnapshot();
  });
  it('can convert pip test result with remediation (pins)', () => {
    const withRemediation = JSON.parse(
      fs.readFileSync(
        path.resolve(
          __dirname,
          '../../../../../',
          'acceptance/fixtures/pip-app-with-remediation/test-graph-results.json',
        ),
        'utf8',
      ),
    );
    const res = convertLegacyTestResultToScanResult(withRemediation);
    expect(res).toMatchSnapshot();
  });
});
