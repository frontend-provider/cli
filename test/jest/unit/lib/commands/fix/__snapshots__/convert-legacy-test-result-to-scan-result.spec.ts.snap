// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert legacy TestResult to ScanResult can convert npm test result with no remediation 1`] = `
Object {
  "facts": Array [],
  "identity": Object {
    "targetFile": "package-lock.json",
    "type": "npm",
  },
  "name": "shallow-goof",
  "policy": "# Snyk (https://snyk.io) policy file, patches or ignores known vulnerabilities.
version: v1.19.0
ignore: {}
patch: {}
",
  "target": Object {},
}
`;

exports[`Convert legacy TestResult to ScanResult can convert npm test result with remediation 1`] = `
Object {
  "facts": Array [],
  "identity": Object {
    "targetFile": "package-lock.json",
    "type": "npm",
  },
  "name": "shallow-goof",
  "policy": "# Snyk (https://snyk.io) policy file, patches or ignores known vulnerabilities.
version: v1.19.0
ignore: {}
patch: {}
",
  "target": Object {},
}
`;

exports[`Convert legacy TestResult to ScanResult can convert pip test result with remediation (pins) 1`] = `
Object {
  "facts": Array [],
  "identity": Object {
    "targetFile": "requirements.txt",
    "type": "pip",
  },
  "name": "pip-app",
  "policy": "# Snyk (https://snyk.io) policy file, patches or ignores known vulnerabilities.
version: v1.19.0
ignore: {}
patch: {}
",
  "target": Object {},
}
`;
